docs/working-copy.md

```diff
@@ -12,12 +12,14 @@ working-copy contents when they have changed. Most `jj` commands you run will
 commit the working-copy changes if they have changed. The resulting revision
 will replace the previous working-copy revision.
 
-Also unlike most other VCSs, added files are implicitly tracked. That means that
-if you add a new file to the working copy, it will be automatically committed
-once you run e.g. `jj st`. Similarly, if you remove a file from the working
-copy, it will implicitly be untracked. To untrack a file while keeping it in
-the working copy, first make sure it's [ignored](#ignored-files) and then run
-`jj untrack <path>`.
+Also unlike most other VCSs, added files are implicitly tracked by default. That
+means that if you add a new file to the working copy, it will be automatically
+committed once you run e.g. `jj st`. Similarly, if you remove a file from the
+working copy, it will implicitly be untracked. The `snapshot.auto-track` config
+option controls which paths get automatically tracked when they're added. See
+the [fileset documentation](filesets.md) for the syntax. To untrack a file while
+keeping it in the working copy, first make sure it's [ignored](#ignored-files)
+and then run `jj untrack <path>`.
```

Comment:
> The content is not lost. 

So, both `track` and `untrack` do a snapshot?

Reply:
Yes. The only thing that's lost is whether the file was tracked. The content is not lost.

Reply:
> I think we should do away with the ambiguous states of tracked/ignored and untracked/not ignored.

Every version control system before `jj` had these states well defined. This does not mean they were a good idea but they were certainly not "ambiguous".

> The consequence would be that you would be reminded of the ignored files by every jj status command.

... or reminded of none. Either way this would get rid of the distinction between "temporary files with random names" versus "build directory with predictable name" which is probably the main use case discussed in #323. @joyously has been repeatedly trying to push back against this distinction in multiple  ways across multiple comments; this is just the latest attempt.
- https://github.com/martinvonz/jj/pull/4338#discussion_r1730438561 (previous comment)
- https://github.com/martinvonz/jj/pull/4338#discussion_r1730392319
- https://github.com/martinvonz/jj/pull/4338#discussion_r1730388156
- https://github.com/martinvonz/jj/issues/323#issuecomment-2308609293
- https://github.com/martinvonz/jj/issues/323#issuecomment-2218074254
- https://github.com/martinvonz/jj/issues/323#issuecomment-2308041154
- https://github.com/martinvonz/jj/issues/323#issuecomment-2308084601

Reply:
I don’t think it’s true that every historical version control system has had Git’s exact tracking/ignore system, but please be civil about the differences of opinion here rather than turning the discussion into a personal dispute…

Reply:
Recommending at least 7 times "don't work like this" in more or less subtle and different ways may look very civil on the surface, and I did assume good faith most of the time: you can check the record. I'm afraid it's not possible to assume good faith anymore. 

I don't know whether `jj` can support the usage we described in #323 without affecting existing users (and I'm extremely grateful to @martinvonz for trying - on the weekend!) If that's not possible then fine: people like me will have to either change or not use `jj`. Disappointing but we'll all survive. My problem is: pretending to help while actually doing the opposite. This is time-consuming, not productive and not respectful.

Please propose a better and more civil way I should express this so I can learn.

Reply:
Two conversations is not "repeatedly trying to push back". I ask questions to explore the consequences that might not have been considered. It really doesn't matter to me which way it goes, but this topic does point out a few ingrained workflows that might benefit from the slight changes that `jj` makes to how a VCS works.

Reply:
> > > I think we should do away with the ambiguous states of tracked/ignored and untracked/not ignored.

> > At least outside jj, tracked + ignored is not so rare. For instance it's pretty common to ignore *.o and yet have a couple of binaries or other generated files tracked - because they were generated outside that project.

> Well, it could be that you keep ignores as is, but make the track and untrack command update the ignore file.

The `.gitignore` format is surprisingly complex: wildcards, negations, precedence rules... Should that complexity be duplicated in `jj`? It's not even just one file: you can have one `.gitignore` file in each subdirectory; which one should be updated? Also, what happens when the user modifies their HOME `~/.gitconfig` file? Or is this about some `jj`-specific ignore file, combined somehow with git's?

Also, when and where do you show untracked == ignored files and which ones? Always? Never? Depending on the `jj` command? Depending on some `jj cmd --ignored` option? A mix of the above? You've simplified the data model which always looks good on paper, but that increased the distance with real-world use cases from #323, which at best complexifies the user interface in order to bridge the gap between the two.


> this topic does point out a few ingrained workflows that might benefit from the slight changes that jj makes to how a VCS works.

Thanks: please be more direct and clearly explicit these "ingrained workflows" and usages that you think should simply not be supported at all rather than demonstrating how to avoid them in #323 and suggesting designs here that make them obviously harder or impossible to support. Then we can just agree to disagree on use cases and move on.

Brainstorming and prototyping is great and useful, but at some point a formally defined, high-level design that lists  what's possible versus not will be required; at the very least for documentation purposes if nothing else. Some back and forth is inevitable but right now I find it quite hard to discuss both implementation details and possibly shifting requirements at the same time.

Reply:
> I don’t think it’s true that every historical version control system has had Git’s exact tracking/ignore system,

I don't know whether they were all similar (probably not) but for sure every system had to have some "unambiguous" definition of those states: defined at least by their respective implementation if not documented formally/properly.  I don't remember track/ignore definitions fundamentally different from the git ones and it looks like these stood the test of time. Recent tools like git have the huge benefit of hindsight (especially from BitKeeper but I digress).

Reply:
> > I think we should do away with the ambiguous states of tracked/ignored and untracked/not ignored.
> 
> Interesting idea! That basically means dropping support for ignores and having just the auto-tracking config.
> > That basically means dropping support for ignores and having just the auto-tracking config.
> 
> Well, it could be that you keep ignores as is, but make the `track` and `untrack` command update the ignore file.

Thanks for bringing it up,  as I really like the idea, @joyously could you expand on it in #3525? While I have not deeply thought about the implications yet, I'd like to try such a approach for a native backend.

Reply:
If we want to show untracked files in `jj status`, then I feel like it would be better to have them be separate concepts, because ignored files shouldn't show up in `jj status`.

Reply:
> If we want to show untracked files in jj status, then I feel like it would be better to have them be separate concepts, because ignored files shouldn't show up in jj status.

I don't think we want that, usually, do we?

There are some exceptional case where we would want these to show up in `jj status`, such as files that weren't tracked because they are too large (see e.g. https://github.com/martinvonz/jj/issues/3912) or perhaps files that aren't tracked because `jj` detected some secrets that shouldn't be committed in them (once `jj` becomes all-too-smart). But in the normal case, if the user is aware these files aren't tracked, I don't think they should show up in `status`.

More generally, this discussion made me follow a few rabbit holes, see https://github.com/martinvonz/jj/pull/4338#discussion_r1741304289 for one of them.

Reply:
Personally, I think it would be better to encourage the user to make a explicit decision regarding untracked files (either to explicitly start tracking them or explicitly mark them as ignored), since that would prevent mistakes in either direction (accidentally committing files or forgetting to commit files). So based on that, I think it would be nice if untracked files showed up in `jj status` as a reminder to the user that they should make a decision.

I also think it would be simpler to make all untracked-but-not-explicitly-ignored files show up in `jj status` than to have special cases for large files or secrets (I could even imagine a future where there are `larger_than(1MiB)` and `contains_secret()` filesets, and these checks are implemented by excluding those filesets from `snapshot.auto-track`).

Another alternative would be to just give a message the first time an untracked file is snapshotted (even though it isn't added to the working copy commit), similar to how `jj git fetch` notifies about new untracked branches.

I could also see an argument that setting `snapshot.auto-track` is already making the final decision not to track any files that aren't included in it, but in that case I feel like the name would be misleading, and it should be called something like `snapshot.ignore` instead.

But I think having any config option at all is nice, and any UX improvements/bikeshedding could be done later.

Reply:
> But I think having any config option at all is nice, and any UX improvements/bikeshedding could be done later.

I generally agree, I wouldn't block this PR on fully resolving these questions.

------------------

I think there are now at least two orthogonal features being discussed. One is to have a category of files that the user is reminded about in `jj status`. I do not very much like the name of "untracked but not ignored files" for this concept, but substantially having such a concept might make a lot of sense. I'd try to make it more like a tri-state: a path could be definitely tracked, definitely untracked, or pending a user decision. In my mind, this is closely related to the question of how to handle the `larger_than(1MiB)` files. 

The previous paragraph's feature would be a different feature than the feature of having a per-repo (or per-workspace) list of ignored files that is not shared with the other people working on the repo. I *thought* the latter is closer to what https://github.com/martinvonz/jj/issues/323 and this PR are about, though I might be wrong.

It's a confusing topic, since there are quite a few orthogonal features that can be thought of as related to ignored/untracked files.

> I could also see an argument that setting snapshot.auto-track is already making the final decision not to track any files that aren't included in it,

Yes, that's what I thought, though (again) I'm not entirely certain whether that is what Martin intended or what most people from #323 were thinking.

> but in that case I feel like the name would be misleading, and it should be called something like snapshot.ignore instead.

It didn't seem misleading to me, but that doesn't necessarily mean it isn't.

Either way, it is a good question of whether it's easier for the user to think in terms of ignore patterns or in terms of the opposite, patterns for files to track. In a vacuum, I'd go with the latter like Martin did, but perhaps people are so used to ignore patterns that the other way is better? I imagine people would get confused if we did something like https://github.com/martinvonz/jj/pull/4338#discussion_r1741304289 in terms of positive patterns rather than ignore patterns.